### PR TITLE
boards: arm: Add Thingy:52

### PR DESCRIPTION
**Pull request by tkln**
Add initial support for Thingy:52 / nrf52_pca20020.

### CODE REVIEW

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+
+nRF52-PCA20020 (Thingy:52)
+##########################
+
+Overview
+########
+
+Zephyr uses the nrf52_pca20020 board configuration for building for the
+Thingy:52 board. The board has the nRF52832 MCU with ARM Cortex-M4F processor,
+a set of environmental sensors, a pushbutton and two RGB LEDs.
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: add a comma: ``pushbutton, and``

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
++-------------------+-------------+
+| 6                 | Blue        |
++-------------------+-------------+
+| 7                 | Red         |
++-------------------+-------------+
+
+Button
+------
+
+Thingy:52 has a pushbutton, it's connected to the P0.11 SOC GPIO pin.
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: delete ``, it's``

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
++---------+---------+
+| SOC Pin | Signal  |
++=========+=========+
+| P0.07   | SDA     |
++---------+---------+
+| P0.08   | SCL     |
++---------+---------+
+
+Following devices are attached to the bus.
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: ``The following devices``

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+| VREG    | The battery  | Always on            |
++---------+--------------+----------------------+
+| VDD_nRF | VREG         | Always on            |
++---------+--------------+----------------------+
+| VDD     | VREG         | SOC pin P0.30        |
++---------+--------------+----------------------+
+| VDD_CCS | VDD          | GPIO expander pin 10 |
++---------+--------------+----------------------+
+
+For a given device to be powered, all of the rails from which device's power
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: How about:

    All power rails used for a device must be powered for a given device to be powered.

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+========
+
+Flashing Zephyr onto Thingy:52 requires an external J-Link programmer. The
+programmer is attached to the P9 programming header.
+
+
+Debugging
+=========
+
+Thingy:52 does not have an on-board J-Link debug IC like some other nRF5
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: change ``like`` to ``as``

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+development boards, however, instructions from the :ref:`nordic_segger` page
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: change to:

    development boards do.  However, ...

**Review by galak — CHANGES_REQUESTED** [review]

File: drivers/gpio/Kconfig.nrf5 [hunk 7]
```diff
@@ -14,6 +14,12 @@ menuconfig GPIO_NRF5
 
 if GPIO_NRF5
 
+config GPIO_NRF5_INIT_PRIORITY
```
Comment (galak) [inline drivers/gpio/Kconfig.nrf5]: Why is this needed, please add a comment in the commit message explaining why we need to configure the init priority

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+also apply to this board, with the additional step of connecting an external
+debugger. A development board with a Debug out connector like the
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: change ``like`` to ``such as``

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+:ref:`nrf52_pca10040` can be used as a debugger with Thingy:52.
+
+Testing board features
+**********************
+
+The green lightwell LED can be tested with the blinky example.
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: change ``blinky`` to `` :ref:`blinky-sample` ``  (this will link to the blinky documentation page)

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/basic/blinky
+   :board: nrf52_pca20020
+   :goals: build flash
+
+
+
+Also the temperature and humidity sensor can be tested with the following
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: change ``following sample`` to `` :ref:`hts221` sample`` (links to the sample's documentation)

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 11]
```diff
@@ -0,0 +1,354 @@
+.. _nrf52_pca20020:
+
+nRF52-PCA20020 (Thingy:52)
+##########################
+
+Overview
+########
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: Change this heading underline to use ``****``'s

File: boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
++-----+---------------+-----------------------+
+
+MOSFETs
+~~~~~~~
+
+The MOSFETs are attached to the following SOC GPIO pins:
+
++----------+----------+
+| Device   | Gate Pin |
++----------+----------+
```
Comment (dbkinder) [inline boards/arm/nrf52_pca20020/doc/nrf52_pca20020.rst]: Use ``====``'s as the row separator here to indicate the previous line is a heading.

**Review by dbkinder — CHANGES_REQUESTED** [review]
some recommended edits

File: drivers/gpio/Kconfig.nrf5 (hunk 7 above)
Comment (tkln) [reply drivers/gpio/Kconfig.nrf5]: This isn't strictly needed, but I think it makes the board power rail init clearer. The issue is that the GPIO controller has to be initialized before the VDD rail can be powered up. Due to this dependency I felt that making the GPIO controller have its own init priority would make sense. I considered not doing this and explaining in the board Kconfig file where the rail initialization priorities are set that the BOARD_VDD_PWR_CTRL_INIT_PRIORITY has to be larger than KERNEL_INIT_PRIORITY_DEFAULT because that's when the GPIO controller is currently initialized. I could drop this commit and do that instead, if you think it's the better option.

**Comment by tkln** [comment]
@galak @dbkinder Fixed. The build failure is unrelated to this PR.

**Review by carlescufi — APPROVED** [review]
LGTM, nice job

**Comment by carlescufi** [comment]
@tkln can you rebase? We've renamed the SoC family and this will need a fix

**Review by dbkinder — APPROVED** [review]
Thanks!

**Comment by tkln** [comment]
@carlescufi Done. @galak Do you request further changes?

**Comment by tkln** [comment]
@galak ?

File: boards/arm/nrf52_pca20020/nrf52_pca20020.dts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+		label = "LPS22HB";
+	};
+
+	hts221: hts221@0x5f {
+		compatible = "st,hts221";
+		reg = <0x5f>;
+		label = "HTS221";
+	};
+
+	ccs811: ccs811@0x5a {
```
Comment (galak) [inline boards/arm/nrf52_pca20020/nrf52_pca20020.dts]: remove 0x should be ccs811@5a

File: boards/arm/nrf52_pca20020/nrf52_pca20020.dts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+		label = "GPIO_P0";
+	};
+
+	lps22hb_press: lps22hb_press@0x5c {
+		compatible = "st,lps22hb-press";
+		reg = <0x5c>;
+		label = "LPS22HB";
+	};
+
+	hts221: hts221@0x5f {
```
Comment (galak) [inline boards/arm/nrf52_pca20020/nrf52_pca20020.dts]: remove 0x, should be hts221@5f

File: boards/arm/nrf52_pca20020/nrf52_pca20020.dts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+	status = "ok";
+	clock-frequency = <I2C_BITRATE_FAST>;
+
+	sx1509b: sx1509b@3e {
+		compatible = "semtech,sx1509b";
+		reg = <0x3e>;
+		label = "GPIO_P0";
+	};
+
+	lps22hb_press: lps22hb_press@0x5c {
```
Comment (galak) [inline boards/arm/nrf52_pca20020/nrf52_pca20020.dts]: remove 0x, should be lps22hb_press@5c

**Review by galak — CHANGES_REQUESTED** [review]
Cleanup 0x in device tree (this will also need changes in dts.fixup file)>

**Comment by tkln** [comment]
@galak Done.

**Review by galak — APPROVED** [review]
